### PR TITLE
Add ERA5 and firn model updates

### PR DESCRIPTION
**Pull request by amschne**
This PR brings in an option to force the land model with ERA5 surface variables. ERA5 is configured with additions to datm.

This PR also includes tweaks to the firn density parameterization and a new cold start condition. The new cold start (initial) condition sets the mass of snow on glaciers and ice sheets to one half of the maximum allotted SWE. The maximum allotted mass of snow in a column is determined by the constant "h2osno_max," which is set to 1 m by default. With "use_extrasnowlayers" active (firn model on), this PR sets h2osno_max to 30 m (SWE).

In this PR, the initial condition for a cold start with use_extrasnowlayers is therefore 15 m SWE, which represents a "deep firn" layer with a density of (~)730 kg / m^3. From this initial condition, a spin up period of at least 300 years is recommended before initializing simulations of nonzero ice sheet SMB.

### CODE REVIEW

File: components/data_comps/datm/cime_config/config_component.xml [hunk 1]
```diff
@@ -10,12 +10,13 @@
        This file may have atm desc entries.
   -->
   <description modifier_mode="1">
-    <desc atm="DATM[%QIA][%WISOQIA][%CRU][%CRUv7][%GSWP3v1][%MOSARTTEST][%NLDAS2][%CPLHIST][%1PT][%NYF][%IAF][%JRA][%JRA-1p5][%JRA-1p4-2018][%JRA-RYF8485][%JRA-RYF9091][%JRA-RYF0304][%CFSv2][%CFSR]"> Data driven ATM </desc>
+    <desc atm="DATM[%QIA][%WISOQIA][%CRU][%CRUv7][%GSWP3v1][%ERA5][%MOSARTTEST][%NLDAS2][%CPLHIST][%1PT][%NYF][%IAF][%JRA][%JRA-1p4-2018][%JRA-RYF8485][%JRA-RYF9091][%JRA-RYF0304]"> Data driven ATM </desc>
```
Comment (darincomeau) [inline components/data_comps/datm/cime_config/config_component.xml:13]: Note that `JRA-`p5` is mistakenly removed here.

File: components/data_comps/datm/cime_config/config_component.xml [hunk 2]
```diff
@@ -43,13 +44,13 @@
 
   <entry id="DATM_MODE">
     <type>char</type>
-    <valid_values>CORE2_NYF,CORE2_IAF,CLM_QIAN,CLM_QIAN_WISO,CLM1PT,CLMCRUNCEP,CLMCRUNCEPv7,CLMGSWP3v1,CLMMOSARTTEST,CLMNLDAS2,CPLHIST,CORE_IAF_JRA,IAF_JRA_1p5,CORE_IAF_JRA_1p4_2018,CORE_RYF8485_JRA,CORE_RYF9091_JRA,CORE_RYF0304_JRA,CFSv2,CFSR</valid_values>
+    <valid_values>CORE2_NYF,CORE2_IAF,CLM_QIAN,CLM_QIAN_WISO,CLM1PT,CLMCRUNCEP,CLMCRUNCEPv7,CLMGSWP3v1,ELMERA5,CLMMOSARTTEST,CLMNLDAS2,CPLHIST,CORE_IAF_JRA,CORE_IAF_JRA_1p4_2018,CORE_RYF8485_JRA,CORE_RYF9091_JRA,CORE_RYF0304_JRA</valid_values>
```
Comment (darincomeau) [inline components/data_comps/datm/cime_config/config_component.xml:47]: here as well - `IAF_JRA_1p5` should remain in.

File: components/data_comps/datm/cime_config/config_component.xml (hunk 1 above)
Comment (amschne) [reply components/data_comps/datm/cime_config/config_component.xml]: Thanks @darincomeau. I added back the mistakenly removed datasets.

File: components/data_comps/datm/cime_config/config_component.xml (hunk 2 above)
Comment (amschne) [reply components/data_comps/datm/cime_config/config_component.xml]: Added back in latest commit

**Comment by amschne** [comment]
This PR needs further testing. Short runs (< 10 years) are working but long runs (> 10 years) are crashing due to a calendar issue. See https://github.com/E3SM-Project/E3SM/pull/4806 for details.

Also needed is a workflow to reformat the raw ERA5 output. I have completed years 1980--1989 with a python script. But if changes are needed to these data and / or 1990-2022 is desired, then more data will have to be processed. We currently do not have access to a machine with enough storage space (~1 TB per year) and memory (>100GB) to complete the process. I've tried processing the data in parallel (1 year per task) on cori but it runs really slow. I have some batch scripts in my home directory on cori `/global/homes/a/amschnei/projects/add_era5_forcing_elm`

**Comment by wlin7** [comment]
Hi @amschne, (hope you are still able to see this message  -- here or the auto-generated email) . Thanks for getting the ERA5 forcing about ready for E3SM use. Sooner or later we will need the data for other years. I would like to obtain a copy of your scripts you mentioned above but got permission denied. Can you please give read permission (or create a copy under /global/cscratch1/sd/wlin/share/Adam.S)? Many thanks.

As for storage need,  we could reduce the temporal frequency (say from hourly to 3 hourly or even 6 hourly) if not having enough space.

**Comment by amschne** [comment]
Thanks for pointing that out, @wlin7. I changed the permissions so that everyone has read access to the above directory.

**Comment by matthewhoffman** [comment]
These changes have been incorporated in https://github.com/E3SM-Project/E3SM/pull/6298 and https://github.com/E3SM-Project/E3SM/pull/6234

**Pull request closed without merging**